internal/client/api/leaves.go

```diff
@@ -0,0 +1,351 @@
+) (LeafSet[H, N], error) {
+	storage := btree.NewMap[N, []H](0)
+
+	leaves := db.Get(database.ColumnID(column), prefix)
+	if leaves != nil {
+		type numberHashes struct {
+			Number N
+			Hashes []H
+		}
+		var vals []numberHashes
+		err := scale.Unmarshal(leaves, &vals)
+		if err != nil {
+			return LeafSet[H, N]{}, err
+		}
+		for _, nh := range vals {
+			storage.Set(nh.Number, nh.Hashes)
+		}
+	}
+	return LeafSet[H, N]{
+		storage: *storage,
+	}, nil
+}
+
+// Import will update the leaf list on import.
+func (ls *LeafSet[H, N]) Import(hash H, number N, parentHash H) ImportOutcome[H, N] {
+	var removed *H
+	if number != 0 {
+		parentNumber := number - 1
+		if ls.removeLeaf(parentNumber, parentHash) {
+			removed = &parentHash
+		}
+	}
+
+	ls.insertLeaf(number, hash)
+	return ImportOutcome[H, N]{
+		inserted: leafSetItem[H, N]{hash, number},
+		removed:  removed,
+	}
+}
+
+// Remove will update the leaf list on removal.
+//
+// Note that the leaves set structure doesn't have the information to decide if the
+// leaf we're removing is the last children of the parent. Follows that this method requires
+// the caller to check this condition and optionally pass the `parentHash` if `hash` is
+// its last child.
+//
+// Returns `nil` if no modifications are applied.
+func (ls *LeafSet[H, N]) Remove(hash H, number N, parentHash *H) *RemoveOutcome[H, N] {
+	if !ls.removeLeaf(number, hash) {
+		return nil
+	}
+
+	var inserted *H
+	if parentHash != nil {
+		if number != 0 {
+			parentNumber := number - 1
+			ls.insertLeaf(parentNumber, *parentHash)
+			inserted = parentHash
+		}
+	}
+
+	return &RemoveOutcome[H, N]{
+		inserted: inserted,
+		removed:  leafSetItem[H, N]{hash, number},
+	}
+}
+
+// FinalizeHeight will note a block height finalized, displacing all leaves with number less than the finalized
+// block's.
+//
+// Although it would be more technically correct to also prune out leaves at the
+// same number as the finalized block, but with different hashes, the current behavior
+// is simpler and our assumptions about how finalization works means that those leaves
+// will be pruned soon afterwards anyway.
+func (ls *LeafSet[H, N]) FinalizeHeight(number N) FinalizationOutcome[H, N] {
+	var boundary N
+	if number == 0 {
+		removed := btree.NewMap[N, []H](0)
+		return FinalizationOutcome[H, N]{removed: *removed}
+	}
+	boundary = number - 1
+	belowBoundary := btree.NewMap[N, []H](0)
+	ls.storage.Ascend(boundary, func(key N, value []H) bool {
+		belowBoundary.Set(key, value)
+		ls.storage.Delete(key)
+		return false
+	})
+	return FinalizationOutcome[H, N]{removed: *belowBoundary}
+}
+
+// DisplacedByFinalHeight is the same as `FinalizeHeight()`, but it only simulates the operation.
+//
+// This means that no changes are done.
+//
+// Returns the leaves that would be displaced by finalizing the given block.
+func (ls *LeafSet[H, N]) DisplacedByFinalHeight(number N) FinalizationOutcome[H, N] {
```

Comment:
Do you know in what context this is used? I see it used in `DisplacedLeavesAfterFinalizing` however that function is not used anywhere

Reply:
`DisplacedLeavesAfterFinalizing` fulfills `blockchain.Backend` interface.  Based on a quick investigation in substrate uses this method to populate the stale heads in finality notifications.

Reply:
Okay, so then we plan to use this for the same purpose in our code in the future?

Reply:
Yup this will be addressed when I do the integration in #3902 depending on if we actually support finality notifications currently in Gossamer.  A quick search in our codebase makes me believe we do not.  So something I definitely will revisit when I pick up the GRANDPA work again after this.